### PR TITLE
[refactor] memory utils

### PR DESCRIPTION
**Pull request by feifeibear**
1. collect CPU mem stats for memory tracer
2. move Stateful Tensor related APIs to directory zero. avoiding circle import.

### CODE REVIEW

File: colossalai/utils/memory_tracer/memstats_collector.py [hunk 1]
```diff
@@ -23,45 +23,71 @@ def reset(self):
...
+        if type == 'cuda':
+            return self._overall_cuda_list
+        elif type == 'cpu':
+            return self._overall_cpu_list
+        else:
+            raise TypeError
 
     @property
-    def non_model_data_cuda(self):
+    def model_data_cuda_list(self, type: str, unit: str = 'B') -> List[int]:
```
Comment (Gy-Lu) [inline colossalai/utils/memory_tracer/memstats_collector.py]: maybe a function can reduce the duplicated codes?

File: colossalai/utils/memory_tracer/memstats_collector.py [hunk 2]
```diff
@@ -23,45 +23,71 @@ def reset(self):
...
+        scale = 1
+        if unit == 'GB':
+            scale = 1e9
+        elif unit == 'MB':
+            scale = 1e6
+        elif unit == 'KB':
+            scale = 1e3
+
+        if type == 'cuda':
+            return [(v1 - v2) / scale for v1, v2 in zip(self._overall_cuda_list, self._model_data_cuda_list)]
```
Comment (Gy-Lu) [inline colossalai/utils/memory_tracer/memstats_collector.py:86]: I am not sure if it is the right way to get the non_model_data memory...

File: colossalai/utils/memory_utils/utils.py [hunk 3]
```diff
@@ -1,29 +1,90 @@
...
+def _get_cpu_memory_info():
+    ps_mem_info = namedtuple("ps_mem_info", ["total", "free", "cached", "buffers", "used"])
+    try:
+        # psutil reads the memory info from /proc/memory_info,
+        # which results in returning the host memory instead of
+        # that of container.
+        # Here we try to read the container memory with method in:
+        # https://stackoverflow.com/a/46213331/5163915
+        mems = {}
+        with open("/sys/fs/cgroup/memory/memory.meminfo", "rb") as f:
```
Comment (Gy-Lu) [inline colossalai/utils/memory_utils/utils.py:22]: Is this path universal for all users?

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/utils/memory_tracer/memstats_collector.py:86]: See #562

File: colossalai/utils/memory_utils/utils.py (hunk 3 above)
Comment (feifeibear) [reply colossalai/utils/memory_utils/utils.py:22]: its about docker

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 2 above)
Comment (Gy-Lu) [reply colossalai/utils/memory_tracer/memstats_collector.py:86]: got it

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 1 above)
Comment (Gy-Lu) [reply colossalai/utils/memory_tracer/memstats_collector.py]: not a big deal. we can polish it later.

**Review by Gy-Lu — APPROVED** [review]

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/utils/memory_tracer/memstats_collector.py]: DONE

**Pull request merged**